core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -423,8 +413,11 @@ private[log] object LogValidator extends Logging {
               if (batch.magic > RecordBatch.MAGIC_VALUE_V0 && toMagic > RecordBatch.MAGIC_VALUE_V0) {
```

Comment:
First, I don't fully understand the different between version 0, 1, and 2. I am concern if this check (`if` statement) it still accurate.

1. Why do we only update the `maxTimestamp` if the version is not v0?
2. Why do we disallow in place assignment of records with invalid offsets if the version is not v0?

For 1, is it because we are only allow to update it in v1 and v2? E.g.
```
if (toMagic >= RecordBatch.MAGIC_VALUE_V1)
        batch.setMaxTimestamp(timestampType, maxTimestamp)
```

For 2. is it because in place assignment is always false for `v0`? E.g.:
```
    if (firstBatch.magic != toMagic || toMagic == RecordBatch.MAGIC_VALUE_V0)
      inPlaceAssignment = false
```

Reply:
For 1), it is because the v0 format has no field for the timestamp. For 2), you are right: we always do in place assignment for v0.